src/main/rx/srxl2.c

```diff
@@ -0,0 +1,558 @@
+
+void srxl2RxWriteData(const void *data, int len)
+{
+    len = MIN(len, (int)sizeof(writeBuffer));
+    memcpy(writeBuffer, data, len);
+    writeBufferIdx = len;
+}
+
+bool srxl2RxInit(const rxConfig_t *rxConfig, rxRuntimeConfig_t *rxRuntimeConfig)
+{
+    static uint16_t channelData[SRXL2_MAX_CHANNELS];
+    for (size_t i = 0; i < SRXL2_MAX_CHANNELS; ++i) {
+        channelData[i] = SRXL2_CHANNEL_CENTER;
+    }
+
+    unitId = rxConfig->srxl2_unit_id;
+    baudRate = rxConfig->srxl2_baud_rate;
+
+    rxRuntimeConfig->channelData = channelData;
+    rxRuntimeConfig->channelCount = SRXL2_MAX_CHANNELS;
+    rxRuntimeConfig->rxRefreshRate = SRXL2_FRAME_PERIOD_US;
+
+    rxRuntimeConfig->rcReadRawFn = srxl2ReadRawRC;
+    rxRuntimeConfig->rcFrameStatusFn = srxl2FrameStatus;
+    rxRuntimeConfig->rcProcessFrameFn = srxl2ProcessFrame;
+
+    const serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_RX_SERIAL);
+    if (!portConfig) {
+        return false;
+    }
+
+    portOptions_e options = SRXL2_PORT_OPTIONS;
+    if (rxConfig->serialrx_inverted) {
+        options |= SERIAL_INVERTED;
+    }
+    if (rxConfig->halfDuplex) {
+        options |= SERIAL_BIDIR;
+    }
+
+    serialPort = openSerialPort(portConfig->identifier, FUNCTION_RX_SERIAL, srxl2DataReceive,
+        NULL, SRXL2_PORT_BAUDRATE_DEFAULT, SRXL2_PORT_MODE, options);
+
+    if (serialPort) {
+        serialPort->idleCallback = srxl2Idle;
+
+        state = ListenForActivity;
+        timeoutTimestamp = micros() + SRXL2_LISTEN_FOR_ACTIVITY_TIMEOUT_US;
+
+        if (rssiSource == RSSI_SOURCE_NONE) {
+            rssiSource = RSSI_SOURCE_RX_PROTOCOL;
+        }
+    }
+
+    return serialPort;
+}
+
+bool srxl2RxIsActive(void)
+{
+    return serialPort;
+}
+
+bool srxl2TelemetryRequested(void)
+{
+    return telemetryRequested;
+}
+
+void srxl2InitializeFrame(sbuf_t *dst)
+{
+    dst->ptr = telemetryFrame;
+    dst->end = ARRAYEND(telemetryFrame);
+
+    sbufWriteU8(dst, SRXL2_ID);
+    sbufWriteU8(dst, TelemetrySensorData);
+    sbufWriteU8(dst, ARRAYLEN(telemetryFrame));
+    sbufWriteU8(dst, busMasterDeviceId);
+}
+
+void srxl2FinalizeFrame(sbuf_t *dst)
+{
+  const uint16_t crc = crc16_ccitt_update(0, telemetryFrame, sbufPtr(dst) - telemetryFrame);
+  sbufWriteU16BigEndian(dst, crc);
+
+  sbufSwitchToReader(dst, telemetryFrame);
+  srxl2RxWriteData(sbufPtr(dst), sbufBytesRemaining(dst));
+  telemetryRequested = false;
+}
+
+void srxl2Bind(void)
+{
+    const size_t length = sizeof(Srxl2BindInfoFrame);
+
+    Srxl2BindInfoFrame bind = {
+        .header = {
+            .id = SRXL2_ID,
+            .packetType = BindInfo,
+            .length = length
+        },
+        .payload = {
+            .request = EnterBindMode,
+            .deviceId = busMasterDeviceId,
+            .bindType = DMSX_11ms,
+            .options = SRXL_BIND_OPT_TELEM_TX_ENABLE | SRXL_BIND_OPT_BIND_TX_ENABLE,
+        }
+    };
+
+    const uint16_t crc = crc16_ccitt_update(0, &bind, length - 2);
```

Comment:
CRC calculation is repeated 3 times in this PR, it may be implemented in srxl2RxWriteData (I assume all frames have CRC)

Reply:
Good idea. Crc calculation for all SRXL2 writes are handled in that function in commit 4